README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ihpdr

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/ihpdr)](https://CRAN.R-project.org/package=ihpdr)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![R-CMD-check](https://github.com/kvasilopoulos/ihpdr/workflows/R-CMD-check/badge.svg)](https://github.com/kvasilopoulos/ihpdr/actions)
[![Codecov test
coverage](https://codecov.io/gh/kvasilopoulos/ihpdr/branch/master/graph/badge.svg)](https://codecov.io/gh/kvasilopoulos/ihpdr?branch=master)
<!-- badges: end -->

The goal of {ihpdr} is to fetch data from the [International House Price
Database](https://www.dallasfed.org/institute/houseprice#tab1), compiled
by the Federal Reserve Bank of Dallas.

## Installation

You can install the released version of {ihpdr} from CRAN with:

``` r
install.packages("ihpdr")
```

Or install the development version from
[GitHub](https://github.com/kvasilopoulos/ihpdr) with:

``` r
# install.packages("devtools")
devtools::install_github("kvasilopoulos/ihpdr")
```

## Example

This is a basic example which shows you how to download the data:

``` r
# Raw Data
ihpdr::ihpd_get()
#> # A tibble: 4,941 x 6
#>    Date       country     hpi  rhpi   pdi  rpdi
#>    <date>     <chr>     <dbl> <dbl> <dbl> <dbl>
#>  1 1975-01-01 Australia  7.60  39.1  14.1  72.3
#>  2 1975-04-01 Australia  7.75  38.5  14.4  71.4
#>  3 1975-07-01 Australia  8.04  38.6  14.7  70.6
#>  4 1975-10-01 Australia  8.29  37.7  15.2  69.5
#>  5 1976-01-01 Australia  8.58  37.9  15.5  69.1
#>  6 1976-04-01 Australia  8.84  38.1  15.9  69.5
#>  7 1976-07-01 Australia  9.07  38.3  17.1  71.2
#>  8 1976-10-01 Australia  9.25  37.9  17.4  71.1
#>  9 1977-01-01 Australia  9.47  37.9  17.7  70.7
#> 10 1977-04-01 Australia  9.67  37.7  18.0  70.1
#> # ... with 4,931 more rows

# Exuberance Indicators ~ bsadf
ihpdr::ihpd_get("bsadf")
#> # A tibble: 18,200 x 6
#>    Date       country   type    lag value  crit
#>    <date>     <chr>     <chr> <dbl> <dbl> <dbl>
#>  1 1975-01-01 Australia rhpi      1    NA    NA
#>  2 1975-04-01 Australia rhpi      1    NA    NA
#>  3 1975-07-01 Australia rhpi      1    NA    NA
#>  4 1975-10-01 Australia rhpi      1    NA    NA
#>  5 1976-01-01 Australia rhpi      1    NA    NA
#>  6 1976-04-01 Australia rhpi      1    NA    NA
#>  7 1976-07-01 Australia rhpi      1    NA    NA
#>  8 1976-10-01 Australia rhpi      1    NA    NA
#>  9 1977-01-01 Australia rhpi      1    NA    NA
#> 10 1977-04-01 Australia rhpi      1    NA    NA
#> # ... with 18,190 more rows

# Get the release dates
ihpdr::ihpd_release_dates()
#>   Last Quarter Included   Data Release Date
#> 2    First quarter 2020     July 6–10, 2020
#> 3   Second quarter 2020   October 5–9, 2020
#> 4    Third quarter 2020 January 11–15, 2021
#> 5   Fourth quarter 2020   April 12–16, 2021
```

## Wrangle & Plot Real House Prices

``` r
library(tidyverse)
```

### Raw Data

``` r
raw_data <- ihpdr::ihpd_get()

ggplot(raw_data, aes(Date, rhpi)) + 
  geom_line(size = 0.7) + 
  scale_x_date(date_labels = "%y") +
  facet_wrap(~country, ncol = 4) 
```

<img src="man/figures/README-rhpi-1.png" width="100%" />

### Exuberance Indicators - Date-stamping

``` r
bsadf_data <- ihpdr::ihpd_get("bsadf")

bsadf_data %>% 
  dplyr::filter(type == "rhpi", lag == 1) %>% 
  ggplot() + 
  geom_line(aes(Date, value), size = 0.7) +
  geom_line(aes(Date, crit), col = "red", size = 0.7) +
  scale_x_date(date_labels = "%y") +
  facet_wrap(~country, ncol = 4) +
  theme_bw()
```

<img src="man/figures/README-exuber-ind-1.png" width="100%" />


NEWS.md
# ihpdr 1.3.0

### New functions:
* `ihpd_countries()`: lists all countries
* `ihpf_versions()`: lists all available versions

# ihpdr 1.2.1

* Minor changes to accommodate the new format in 2020-Q1 from the IHPD.

# ihpdr 1.2.0

* `ihpd_get()`: now prints the url of the file that is accessing. It can be opted 
out by `verbose = FALSE`.
* `ihpd_get()` has the option to fetch an older version. Note that prior to 
`1201` (2012: Q1) the files are backward compatible, thus output has different format.
* Accessing internet files fail errors have been demoted to messages to comply with
CRAN guidelines and gracefully fail.


# ihpdr 1.1.0

* `ihpd_browse()`: new function that takes you to various webpages associated with international house price database.
* Bug Fix: Replaced numerical indexing to regexp for excel file selection.

# ihpdr 1.0.0.9000

* Added a `NEWS.md` file to track changes to the package.

# ihpdr 1.0.0

* CRAN Release


tests/testthat.R
library(testthat)
library(ihpdr)

test_check("ihpdr")


tests/testthat/helper.R
skip_if_http_error <- function() {
  remote_file <- "https://www.dallasfed.org/institute/houseprice#tab2"
  skip_if(httr::http_error(remote_file))
}


tests/testthat/test-get.R
context("ihpd_get")

skip_if_offline()
skip_if_http_error()

test_that("get works", {
  expect_error(ihpd_get("raw"), NA)
  expect_error(ihpd_get("gsadf"), NA)
  expect_error(ihpd_get("bsadf"), NA)
})

test_that("older versions work", {
  expect_error(ihpd_get("raw", version = "1704"), NA)
  expect_error(ihpd_get("gsadf", version = "1704"), NA)
  expect_error(ihpd_get("bsadf", version = "1704"), NA)
})

test_that("wrong-version", {
  expect_error(ihpd_get("raw", version = "10-1"))
})


test_that("bsadf has the right elements",{
  bsadf <- ihpd_get("bsadf")
  countries <- unique(bsadf$country)
  lags <- unique(bsadf$lag)
  start_date <- bsadf$Date[1]
  expect_false(all(startsWith(countries, "Aggregate")))
  expect_true(all(lags %in% c(1,4)))
  expect_true(start_date == "1975-01-01")
})

test_that("bsadf has the right elements",{
  gsadf <- ihpd_get("gsadf")
  countries <- unique(gsadf$country)
  lags <- unique(gsadf$lag)
  sig <- unique(gsadf$sig)
  expect_false(all(startsWith(countries, "Aggregate")))
  expect_true(all(lags %in% c(1,4)))
  expect_true(all(sig %in% c(0.90, 0.95, 0.99)))
})


R/tf.R
try_GET <- function(x, ...) {
  tryCatch(
    GET(url = x, timeout(10), ...),
    error = function(e) conditionMessage(e),
    warning = function(w) conditionMessage(w)
  )
}
is_response <- function(x) {
  class(x) == "response"
}

ihpd_files <- function(regex = "hp[0-9]") {
  if (!curl::has_internet()) {
    message("No internet connection.")
    return(invisible(NULL))
  }
  remote <- "https://www.dallasfed.org/institute/houseprice#tab2"
  resp <- try_GET(remote)
  if (!is_response(resp)) {
    message(resp)
    return(invisible(NULL))
  }
  if (httr::http_error(resp)) { # network is down = message (not an error anymore)
    httr::message_for_status(resp)
    return(invisible(NULL))
  }
  resp

  xml2::read_html(resp) %>%
    rvest::html_nodes("a") %>%
    rvest::html_attr("href") %>%
    grep(".xlsx$", ., value = TRUE) %>%
    grep(regex, ., value = TRUE)

}

# GET File ----------------------------------------------------------------

#' @importFrom httr GET write_disk timeout
#' @importFrom rvest html_nodes html_attrs html_attr
#' @importFrom xml2 read_html
#' @importFrom rlang %||%
ihpd_tf <- function(version = NULL, regex = "hp[0-9]", access_info = FALSE){

  search_url <- ihpd_files(regex = regex)

  if (is.null(version)) {
    relative_url <- search_url[1]
  } else {
    type <- gsub("\\[0-9]", "", regex)
    versions <- gsub(paste0(".*",type, "(.+)",".xlsx"), "\\1", search_url)
    if (!version %in% versions) {
      versions_mgs <- paste0(sQuote(versions, FALSE), collapse = " ")
      error_msg <- paste("wrong version number should be one of", versions_mgs)
      stop(error_msg, call. = FALSE)
    }
    relative_url <- grep(version, search_url, value = TRUE)
  }

  url <- paste0("https://www.dallasfed.org", relative_url)

  if (interactive() && access_info) {
    message("Accessing ", url)
  }
  tf <- tempfile(fileext = ".xlsx")
  resp_file <- try_GET(url, write_disk(tf))
  if (!is_response(resp_file)) {
    message(resp_file)
    return(invisible(NULL))
  }
  structure(tf, source = url, class = "access_url")
}

enclose <- function(x) {
  paste0("[", x, "]")
}

print.access_url <- function(x) {
  cat(enclose("Local file"), "\n\t", x, "\n")
  cat(enclose("Remote file"), "\n\t", attr(x, "source"))
}


R/browse.R

#' Quickly browse to important links
#'
#' This function take you to webpages associated with the International House
#' Price Database and returns the url invisibly. Option `app` take you the
#' International Housing Observatory, while `info` take you to Dallas Fed
#' info page of the data.
#'
#' @param pick whether to browse the shiny `info` or the `app` page. Defaults at
#' `info`.
#'
#' @export
#'
#' @examples
#'
#' ihpd_browse()
ihpd_browse <- function(pick = c("info", "app")) {
  pick <- match.arg(pick)
  view_url(pick_url(pick))
}

pick_url <- function(pick) {
  ihpd_url <- switch(
    pick,
    app = "https://lancs-macro.shinyapps.io/international-housing-observatory/",
    info =  "https://www.dallasfed.org/institute/houseprice#tab2"
  )
}

view_url <- function(ihpd_url, open = interactive()) {
  if (open) {
    utils::browseURL(ihpd_url)
  }
  invisible(ihpd_url)
}


R/scrape.R
#' @importFrom dplyr mutate full_join
#' @importFrom tidyr gather
#' @importFrom lubridate yq
#' @importFrom rlang !! enquo
clean_gather <- function(x, nm, ...) {
  x %>% dplyr::rename(Date = 1) %>%
    dplyr::mutate(Date = Date %>% lubridate::yq()) %>%
    tidyr::gather(country, !!enquo(nm), -Date, ...)
}

#' Download International House Price Database Data
#'
#' All releases of the database include the following time series for the
#' countries covered:
#' \itemize{
#'  \item The house price index (hpi).
#'  \item The house price index expressed in real terms (rhpi).
#'  \item The personal disposable income index (pdi).
#'  \item The personal disposable income expressed in real terms index (rpdi).
#' }
#' All variables in real terms are deflated with the personal consumption
#' expenditure (PCE) deflator.
#'
#' @importFrom httr GET write_disk
#' @importFrom rvest html_nodes html_attrs
#' @importFrom xml2 read_html
#' @importFrom readxl read_excel
#' @importFrom tidyr drop_na
#' @export
#'
#' @examples
#' download_raw()
#'
download_raw <- function() {

  relative_url <-
    xml2::read_html("https://www.dallasfed.org/institute/houseprice#tab2") %>%
    rvest::html_nodes("a") %>%
    rvest::html_attr("href") %>%
    grep(".xlsx$", ., value = TRUE) %>%
    `[`(1)

  absolute_url <- paste0("https://www.dallasfed.org", relative_url)

  tf <- tempfile(fileext = ".xlsx")
  httr::GET(absolute_url, write_disk(tf))

  suppressMessages({
    hpi <- readxl::read_excel(tf, sheet = 2)
    rhpi <- readxl::read_excel(tf, sheet = 3)
    pdi <- readxl::read_excel(tf, sheet = 4)
    rpdi <- readxl::read_excel(tf, sheet = 5)
  })

  list(
    clean_gather(hpi, hpi),
    clean_gather(rhpi, rhpi),
    clean_gather(pdi, pdi),
    clean_gather(rpdi, rpdi)) %>%
    reduce(full_join, by = c("Date", "country")) %>%
    drop_na()

}

#' Download International House Price Database Exuberance Data
#'
#' The exuberance indicators files include test statistics for explosive behavior
#' on the real house price index (rhpi) and the ratio of rhpi to real personal
#' disposable incomde (rpdi) at a lag length of 1 and 4, along
#' with the corresponding 95 percent critical values.
#'
#' @param option whether to download the 'gsadf' or 'bsadf' statistic.
#'
#' @importFrom purrr reduce
#' @importFrom dplyr slice rename select mutate_at vars
#' @export
#'
#' @examples
#'
#' download_exuber(option = "gsadf")
#'
#' download_exuber(option = "bsadf")
#'
download_exuber <- function(option = c("gsadf", "bsadf")) {

  option <- match.arg(option)

  relative_url <-
    xml2::read_html("https://www.dallasfed.org/institute/houseprice#tab2") %>%
    rvest::html_nodes("a") %>%
    rvest::html_attr("href") %>%
    grep(".xlsx$", ., value = TRUE) %>%
    `[`(2)

  absolute_url <- paste0("https://www.dallasfed.org", relative_url)

  tf <- tempfile(fileext = ".xlsx")
  httr::GET(absolute_url, httr::write_disk(tf))

  nms <- readxl::read_excel(tf, sheet = 2, range = "G2:AE2") %>%
    dplyr::rename(Date = DATE, crit = "95% CRITICAL VALUES") %>%
    names()

  suppressMessages({
    lag1 <- readxl::read_excel(tf, sheet = 2, skip = 1, na = c("", "NA"))
    lag4 <- readxl::read_excel(tf, sheet = 3, skip = 1, na = c("", "NA"))
  })

  if (option == "bsadf") {
    out <-
      list(
        lag1 %>%
          select(7:31) %>%
          slice(-1) %>%
          set_names(nms) %>%
          clean_gather("rhpi", -crit),
        lag1 %>%
          select(33:57) %>%
          slice(-1) %>%
          set_names(nms) %>%
          clean_gather("ratio", -crit) %>%
          mutate(lag = 1) ) %>%
      reduce(full_join, by = c("Date", "crit", "country")) %>%
      gather(type, value, -Date, -country, -crit, -lag) %>%
      full_join(
        list(
          lag4 %>%
            select(7:31) %>%
            slice(-1) %>%
            set_names(nms) %>%
            clean_gather("rhpi", -crit),
          lag4 %>%
            select(33:57) %>%
            slice(-1) %>%
            set_names(nms) %>%
            clean_gather("ratio", -crit) %>%
            mutate(lag = 4) ) %>%
          reduce(full_join, by = c("Date", "crit", "country")) %>%
          gather(type, value, -Date, -country, -crit, -lag),
        by = c("country", "Date", "crit", "type", "lag", "value")) %>%
      select(Date, country, type, lag, value, crit)
  }else{

    cv <- lag1[3:5, 1:3] %>%
      set_names("sig", "sadf", "gsadf") %>%
      gather(tstat, crit, -sig)
    cv_seq_sadf_lag1 <- lag1[8:30, 1:3] %>%
      set_names("country", "sadf", "gsadf") %>%
      mutate(type = "rhpi", lag = 1) %>%
      mutate_at(vars(sadf, gsadf), as.numeric) %>%
      gather(tstat, value, -country, -type, -lag)
    cv_seq_gsadf_lag1 <- lag1[8:30, c(1,4:5)] %>%
      set_names("country", "sadf", "gsadf") %>%
      mutate(type = "ratio", lag = 1) %>%
      mutate_at(vars(sadf, gsadf), as.numeric) %>%
      gather(tstat, value, -country, -type, -lag)
    cv_seq_sadf_lag4 <- lag4[8:30, 1:3] %>%
      set_names("country", "sadf", "gsadf") %>%
      mutate(type = "rhpi", lag = 4) %>%
      mutate_at(vars(sadf, gsadf), as.numeric) %>%
      gather(tstat, value, -country, -type, -lag)
    cv_seq_gsadf_lag4 <- lag4[8:30, c(1,4:5)] %>%
      set_names("country", "sadf", "gsadf") %>%
      mutate(type = "ratio", lag = 4) %>%
      mutate_at(vars(sadf, gsadf), as.numeric) %>%
      gather(tstat, value, -country, -type, -lag)

    suppressMessages({
      out <- list(cv, cv_seq_sadf_lag1, cv_seq_gsadf_lag1,
                  cv_seq_sadf_lag4, cv_seq_sadf_lag4) %>%
        reduce(full_join) %>%
        select(country, type, tstat, lag, sig, value, crit) %>%
        mutate_at(vars(crit,sig), as.numeric)
    })
  }
  out
}

#' Fetches the latest release dates
#'
#'@export
#'@importFrom rvest html_table
#'@importFrom rlang set_names
release_dates <- function() {

  tbl <- xml2::read_html("https://www.dallasfed.org/institute/houseprice") %>%
    rvest::html_nodes("table") %>%
    rvest::html_table()

  tbl[[1]][-1, ] %>%
    set_names("Last Quarter Included", "Data Release Date")
}


R/get.R


set_source <- function(tbl, url) {
  attr(tbl, "source") <- attr(url, "source")
  tbl
}

set_version <- function(tbl, vers) {
  attr(tbl, "version") <- attr(vers, "version")
  tbl
}


# Raw ---------------------------------------------------------------------


#' @importFrom dplyr mutate full_join
#' @importFrom lubridate yq
#' @importFrom rlang !! enquo
format_excel_raw <- function(x, sheet_num, nm, ...) {
  suppressMessages({
    read_excel(x, sheet = sheet_num) %>%
      dplyr::rename(Date = 1) %>%
      dplyr::mutate(Date = Date %>% lubridate::yq()) %>%
      tidyr::gather(country, !!enquo(nm), -Date, ...)
  })
}

#' @importFrom readxl read_excel
#' @importFrom tidyr drop_na gather
ihpd_get_raw <- function(.version, .access_info) {

  tf <- ihpd_tf(version = .version, regex = "hp[0-9]", access_info = .access_info)
  tf %||% return(invisible(NULL))
  on.exit(file.remove(tf))

  list(
    format_excel_raw(tf, 2, "hpi"),
    format_excel_raw(tf, 3, "rhpi"),
    format_excel_raw(tf, 4, "pdi"),
    format_excel_raw(tf, 5, "rpdi")) %>%
    reduce(full_join, by = c("Date", "country")) %>%
    drop_na() %>%
    set_source(tf) %>%
    set_version(.version)
}

# Bsadf -------------------------------------------------------------------


format_excel_bsadf <- function(x, nm, nms, ...) {
  x %>% slice(-1) %>% set_names(nms) %>%
    dplyr::rename(Date = 1) %>%
    dplyr::mutate(Date = Date %>% lubridate::yq()) %>%
    tidyr::gather(country, !!enquo(nm), -Date, -crit, ...)
}


ihpd_get_bsadf <- function(.version, .access_info) {

  tf <- ihpd_tf(version = .version, regex = "hpta[0-9]", access_info = .access_info)
  tf %||% return(invisible(NULL))
  on.exit(file.remove(tf))
  # tf <- "inst/hpta2001.xlsx"

  nms <- read_excel(tf, sheet = 2, range = "G2:BQ2", .name_repair = "minimal") %>%
    names() %>% unique()
  nms[1:2] <- c("Date", "crit")
  nms <- nms[nms != ""]

  suppressMessages({
    lag1 <- readxl::read_excel(tf, sheet = 2, skip = 1, na = c("", "NA"))[,-c(1:6)]
    lag4 <- readxl::read_excel(tf, sheet = 3, skip = 1, na = c("", "NA"))[,-c(1:6)]
  })

  split_at <- which(colSums(is.na(lag1)) == nrow(lag1))

  tbl_lag1 <-
    list(
      format_excel_bsadf(lag1[, 1:(split_at - 1)], nm = "rhpi", nms),
      format_excel_bsadf(lag1[, (split_at + 1):ncol(lag1)], nm = "ratio", nms)
      ) %>%
    reduce(full_join, by = c("Date", "crit", "country")) %>%
    mutate(lag = 1) %>%
    gather(type, value, -Date, -country, -crit, -lag)

  tbl_lag4 <-
    list(
      format_excel_bsadf(lag4[, 1:(split_at - 1)], nm = "rhpi", nms),
      format_excel_bsadf(lag4[, (split_at + 1):ncol(lag1)], nm = "ratio", nms)
    ) %>%
    reduce(full_join, by = c("Date", "crit", "country")) %>%
    mutate(lag = 4) %>%
    gather(type, value, -Date, -country, -crit, -lag)

   full_join(tbl_lag1, tbl_lag4,
             by = c("country", "Date", "crit", "type", "lag", "value")) %>%
     select(Date, country, type, lag, value, crit) %>%
     set_source(tf) %>%
     set_version(.version)
}

# GSADF -------------------------------------------------------------------

format_excel_gsadf <- function(x, nm, lag = 1) {
  x %>%
    set_names("country", "sadf", "gsadf") %>%
    mutate(type = nm, lag = lag) %>%
    mutate_at(vars(sadf, gsadf), as.numeric) %>%
    gather(tstat, value, -country, -type, -lag)
}

ihpd_get_gsadf <- function(.version, .access_info) {

  tf <- ihpd_tf(version = .version, regex = "hpta[0-9]", access_info = .access_info)
  tf %||% return(invisible(NULL))
  on.exit(file.remove(tf))

  nms <- readxl::read_excel(tf, sheet = 2, range = "G2:AE2") %>%
    dplyr::rename(Date = DATE, crit = "95% CRITICAL VALUES") %>%
    names()

  suppressMessages({
    lag1 <- readxl::read_excel(tf, sheet = 2, skip = 1, na = c("", "NA"))
    lag4 <- readxl::read_excel(tf, sheet = 3, skip = 1, na = c("", "NA"))
  })

  cv <- lag1[3:5, 1:3] %>%
    set_names("sig", "sadf", "gsadf") %>%
    gather(tstat, crit, -sig)

  list(
    format_excel_gsadf(lag1[8:30, 1:3], "rhpi", lag = 1),
    format_excel_gsadf(lag1[8:30, c(1,4:5)], "ratio", lag = 1),
    format_excel_gsadf(lag4[8:30, 1:3], "rhpi", lag = 1),
    format_excel_gsadf(lag4[8:30, c(1,4:5)], "ratio", lag = 1)
       ) %>%
    reduce(full_join, by = c("tstat", "country", "type", "lag", "value")) %>%
    full_join(cv, by = "tstat") %>%
    mutate_at(vars(crit, sig), as.numeric) %>%
    select(country, type, tstat, lag, value, crit, sig) %>%
    set_source(tf) %>%
    set_version(.version)
}

# GET ---------------------------------------------------------------------


#' Download Data from the International House Price Database
#'
#' Available downloads:
#' \enumerate{
#'   \item raw: Raw Data
#'   \item gsadf: Generalized sup ADF statistics
#'   \item bsadf: Backward sup ADF statistic sequence
#'  }
#'
#' @param symbol Which dataset to download.
#' @param version Which version to download. Version number should be a character
#' of the following format %Y%q (e.g. '1801' - corresponds to year 2018,
#' Quarter 1). Versions start from '1102'. Defaults at the latest available.
#' @param verbose whether to print the url of the excel file that is accessing.
#'
#' @details
#'
#' \describe{
#'   \item{raw}{
#'
#'Includes the following time series for the countries covered:
#' \itemize{
#'  \item hpi: The house price index.
#'  \item rhpi: The house price index expressed in real terms.
#'  \item pdi: The personal disposable income index.
#'  \item rpdi: The personal disposable income expressed in real terms index.
#' }
#' All variables in real terms are deflated with the personal consumption
#' expenditure (PCE) deflator.}
#' }
#'
#'The exuberance indicators include
#'\describe{
#' \item{gsadf}{
#'   \itemize{
#'     \item type: real house price index (rhpi) and the ratio of rhpi to real
#'     personal disposable income (rpdi)
#'     \item lag: lag length of 1 and 4
#'     \item sig: significance level for the critical value
#'     \item value: test statistics for explosive behavior
#'     \item crit: 95 percent critical values.
#'  }
#' }
#'
#' \item{bsadf}{
#'   The exuberance indicators include
#'   \itemize{
#'     \item type: real house price index (rhpi) and the ratio of rhpi to real
#'     personal disposable income (rpdi)
#'     \item lag: lag length of 1 and 4,
#'     \item value: test statistics for explosive behavior
#'     \item crit: 95 percent critical values.
#'     }
#'  }
#' }
#'
#' @importFrom purrr reduce
#' @importFrom dplyr slice rename select mutate_at vars
#'
#' @return Returns a tibble in long format.
#' @export
#'
#' @examples
#'
#' ihpd_get()
ihpd_get <- function(symbol = c("raw", "gsadf", "bsadf"), version = NULL,
                     verbose = TRUE) {
  symbol <- match.arg(symbol)
  switch(
    symbol,
    raw = ihpd_get_raw(.version = version, .access_info = verbose),
    gsadf = ihpd_get_gsadf(.version = version, .access_info = verbose),
    bsadf = ihpd_get_bsadf(.version = version, .access_info = verbose)
  )
}


#' Fetches the latest release dates
#'
#' @export
#' @importFrom rvest html_table
#' @importFrom rlang set_names
#'
#' @return A data.frame with 2 variables with the latest release dates data.
#'
#' @examples
#' ihpd_release_dates()
ihpd_release_dates <- function() {

  tbl <- xml2::read_html("https://www.dallasfed.org/institute/houseprice") %>%
    rvest::html_nodes("table") %>%
    rvest::html_table()

  tbl[[1]][-1, ] %>%
    set_names("Last Quarter Included", "Data Release Date")
}

#' Get all the available versions
#'
#' @export
#' @return A character vector with all the available versions.
#' @examples
#' ihpd_versions()
#'
ihpd_versions <- function() {
  search_url <- ihpd_files(regex = "hp[0-9]")
  type <- gsub("\\[0-9]", "", "hp[0-9]")
  gsub(paste0(".*",type, "(.+)",".xlsx"), "\\1", search_url)
}

#' List all the countries
#'
#' @inheritParams ihpd_get
#' @export
#' @return A character vector all the countries.
#' @examples
#' ihpd_countries()
#'
ihpd_countries <- function(version = NULL) {
  unique(ihpd_get("raw", version = version, verbose = FALSE)$country)
}
